sklearn/preprocessing/resample.py

```diff
@@ -0,0 +1,263 @@
+    array([2, 1, 2])
+    """
+    if n_classes > n_samples:
+        raise ValueError("The number of classes is greater"
+                         " than the number of samples requested")
+    sample_size = n_samples // n_classes
+    sample_size_rem = n_samples % n_classes
+    counts = np.repeat(sample_size, n_classes)
+    if sample_size_rem > 0:
+        counts[:sample_size_rem] += 1
+        # Shuffle so the class inbalance varies between runs
+        random_state = check_random_state(random_state)
+        random_state.shuffle(counts)
+    return counts
+
+
+def _scale_n_samples(scaling, n):
+    """Helper function to scale the number of samples."""
+    if scaling is None:
+        return n
+    else:
+        if isinstance(scaling, numbers.Number) and scaling < 0:
+            raise ValueError("Scaling must be nonnegative: %s" % scaling)
+        elif isinstance(scaling, float):
+            return scaling * n
+        elif isinstance(scaling, (numbers.Integral, np.integer)):
+            return scaling
+        else:
+            raise ValueError("Invalid value for scaling, must be "
+                             "float, int, or None: %s" % scaling)
+
+
+def _weighted_sample(probas, n_samples, random_state=None):
+    """Select indices from n_samples with a weighted probability.
+
+    Parameters
+    ---------
+    probas : array-like
+        Array of probabilities summing to 1.
+    n_samples : integer
+        The number of samples to draw from at random.
+    random_state : int, or RandomState instance (optional)
+        Control the sampling for reproducible behavior.
+    """
+    random_state = check_random_state(random_state)
+    if abs(sum(probas) - 1.0) > .011:
+        raise ValueError("Label distribution probabilites must sum to 1")
+    cum_probas = np.cumsum(probas)
+    cum_probas[-1] = 1  # ensure that the probabilities sum to 1
+    space = np.linspace(0, 1, 10000)
+    weighted_indices = np.searchsorted(cum_probas, space)
+    rs = random_state.rand(n_samples)
+    rs *= len(space)
+    return weighted_indices[rs.astype(int)]
+
+
+def resample_labels(y, method=None, scaling=None, replace=False,
+                    shuffle=False, random_state=None):
+    """Resamples a classes array `y` and returns an array of indices
+    The default behavior is to output the same ``y``. The additional
+    parameters control the desired class distribution of the indices and the
+    number of samples in the output.
+
+    Parameters
+    ----------
+    y : array-like of shape [n_samples]
+        Target classes. Pass in the entire classes array so that
+        that this function can work on the class distribution.
+    method : "balance", "oversample", "undersample", dict (optional)
+        None outputs samples with the same class distribution as `y`.
+        "balance" rebalances the classes to be equally distributed,
+            over/undersampling for `len(y)` samples by default.
```

Comment:
I wanted to indent it some so it doesn't run together with the next line. How about two spaces instead?